### PR TITLE
 Fallback to local repo on errors when proxy to remote repo

### LINKED ISSUE #12853: failed to pull image by proxy-cache after change /etc/hosts
build 3833
1. docker pull busybox by proxy cache
2. change /etc/hosts of harbor-core
image
3. docker pull busybox by proxy cache again, it will fail with 500 Internal Server Error
image
4. log in harbor-core show timeout error
image

This does not meet the description of test case

### PR DESCRIPTION
**Pull request by stonezdj**
When the remote repo is offline or network issue or credential issue, fall back to local repo.
 Fixes #12853

Signed-off-by: stonezdj <stonezdj@gmail.com>

### CODE REVIEW

File: src/server/middleware/repoproxy/proxy.go [hunk 1]
```diff
@@ -113,19 +113,27 @@ func handleManifest(w http.ResponseWriter, r *http.Request, next http.Handler) e
 	log.Debugf("the tag is %v, digest is %v", art.Tag, art.Digest)
 	man, err := proxyCtl.ProxyManifest(ctx, p, art)
 	if err != nil {
-		return err
+		fallbackToLocal(w, r, next, err)
```
Comment (ywk253100) [inline src/server/middleware/repoproxy/proxy.go:116]: Should we check whether the error is not found?

**Comment by ywk253100** [comment]
Is this a reasonable solution to fallback to local when cannot connect to the upstream registry no matter what the error is? This will cause the pulled artifacts out of sync with the upstream ones.

The timeliness is important for the proxy cache? If it is not, users can just push the artifacts into harbor directly and pull the fixed artifacts from harbor directly

File: src/server/middleware/repoproxy/proxy.go (hunk 1 above)
Comment (stonezdj) [reply src/server/middleware/repoproxy/proxy.go]: OK, return the NotFoundError

**Review by reasonerjt — APPROVED** [review]

**Comment by reasonerjt** [comment]
> Is this a reasonable solution to fallback to local when cannot connect to the upstream registry no matter what the error is? This will cause the pulled artifacts out of sync with the upstream ones.
> 
> The timeliness is important for the proxy cache? If it is not, users can just push the artifacts into harbor directly and pull the fixed artifacts from harbor directly

IMO this is a expected behavior of a cache, since we call it cache if it failed to get latest data and it is not expired, it should be served to client

**Review by ywk253100 — APPROVED** [review]

**Pull request merged**